cli/src/commands/git/fetch.rs

```diff
@@ -108,16 +108,25 @@ pub fn cmd_git_fetch(
         }
     }
 
-    match &unmatched_patterns[..] {
-        [] => {} // Everything matched, all good
-        [pattern] if pattern.is_exact() => {
-            return Err(user_error(format!("No git remote named '{pattern}'")))
-        }
-        patterns => {
-            return Err(user_error(format!(
-                "No matching git remotes for patterns: {}",
-                patterns.iter().join(", ")
-            )))
+    match (!matching_remotes.is_empty(), &unmatched_patterns[..]) {
+        (_, []) => {} // Everything matched, all good
+        (has_matches, patterns) => {
+            let msg = match patterns {
+                [pattern] if pattern.is_exact() => {
+                    format!("No git remote named '{pattern}'")
+                }
+                patterns => {
+                    format!(
+                        "No matching git remotes for patterns: '{}'",
+                        patterns.iter().join("', '")
+                    )
+                }
+            };
+            if has_matches {
+                writeln!(ui.warning_default(), "{msg}; continuing with other remotes")?;
+            } else {
+                return Err(user_error(msg));
+            }
```

Comment:
Since unmatched pattern is no longer a hard error, it's probably better to move the warning to the previous loop, and print error/status message if no match found:
```rs
    for pattern in remote_patterns {
        ...
        if remotes.is_empty() {
            // "No matching remotes" warning
        } else {
            matching_remotes.extend(remotes);
        }
    }

    if matching_remotes.is_empty() {
        // "No remotes to push"
        // I'm not sure if this should be an error because pushing to
        // 0 remotes can be considered a valid operation.
    }

Reply:
Yeah, this is probably quite a bit nicer - I'll take a stab at that.

> // I'm not sure if this should be an error because pushing to
> // 0 remotes can be considered a valid operation.

I only considered it because the current behavior is to error in this case but fair point, if you set:
```toml
[git]
fetch = []
```
and run `jj git fetch`, there is no warning about empty remotes, only a branch warning:
```
Warning: No branch matching `*` found on any specified/configured remote
Nothing changed.
```

Reply:
FYI: I kept it an error if there are no matching remotes. This avoids the `Warning: No branch matching '*' found on any specified/configured remote`, which would otherwise print after the more relevant `No git remotes to push.` warning.